src/optimizer/remove_unused_columns.cpp

```diff
@@ -295,66 +480,89 @@ void RemoveUnusedColumns::RemoveColumnsFromLogicalGet(LogicalGet &get) {
+	//! FIXME: pushdown extract is disabled when a struct field is referenced by a filter,
+	//! because that would involve rewriting the existing TableFilterSet
+	SetMode(BaseColumnPrunerMode::DISABLE_PUSHDOWN_EXTRACT);
+
 	// for every table filter, push a column binding into the column references map to prevent the column from
 	// being projected out
-	for (auto &filter : get.table_filters.filters) {
-		optional_idx index;
-		for (idx_t i = 0; i < final_column_ids.size(); i++) {
-			if (final_column_ids[i].GetPrimaryIndex() == filter.first) {
-				index = i;
-				break;
-			}
-		}
-		if (!index.IsValid()) {
-			throw InternalException("Could not find column index for table filter");
-		}
 
+	//! NOTE: This vector is required to keep the referenced Expressions alive
+	vector<unique_ptr<Expression>> filter_expressions;
+	for (auto &filter : get.table_filters.filters) {
+		auto index = GetColumnIdsIndexForFilter(state.old_column_ids, filter.first);
 		auto column_type = get.GetColumnType(ColumnIndex(filter.first));
 
-		ColumnBinding filter_binding(get.table_index, index.GetIndex());
+		ColumnBinding filter_binding(get.table_index, index);
 		auto column_ref = make_uniq<BoundColumnRefExpression>(std::move(column_type), filter_binding);
+		//! Convert the filter to an expression, so we can visit it
 		auto filter_expr = filter.second->ToExpression(*column_ref);
 		if (filter_expr->IsScalar()) {
 			filter_expr = std::move(column_ref);
 		}
-		VisitExpression(&filter_expr);
 		filter_expressions.push_back(std::move(filter_expr));
+		//! Now visit the filter to add to the 'column_references'
+		VisitExpression(&filter_expressions.back());
 	}
 
+	//! Check with the LogicalGet whether pushdown-extract is supported
+	CheckPushdownExtract(get);
+
 	// Clear unused ids, include filter columns that are projected out immediately
 	ClearUnusedExpressions(col_sel, get.table_index);
 
 	// Now set the column ids in the LogicalGet using the "selection vector"
-	vector<ColumnIndex> column_ids;
-	column_ids.reserve(col_sel.size());
 	for (auto col_sel_idx : col_sel) {
+		auto &column_type = get.types[col_sel_idx];
 		auto entry = column_references.find(ColumnBinding(get.table_index, col_sel_idx));
 		if (entry == column_references.end()) {
 			throw InternalException("RemoveUnusedColumns - could not find referenced column");
 		}
-		ColumnIndex new_index(final_column_ids[col_sel_idx].GetPrimaryIndex(), entry->second.child_columns);
-		column_ids.emplace_back(new_index);
+		if (entry->second.child_columns.empty()) {
+			auto logical_column_index = state.old_column_ids[col_sel_idx].GetPrimaryIndex();
+			ColumnIndex new_index(logical_column_index, {});
+			state.AddColumn(col_sel_idx, std::move(new_index));
+			continue;
+		}
+		//! Only a subset of the struct's fields are referenced
+		auto logical_column_index = state.old_column_ids[col_sel_idx].GetPrimaryIndex();
+		if (!entry->second.supports_pushdown_extract) {
```

Comment:
FIXME: this is 100% the same as the case where `child_columns` is empty, unify the two